### PR TITLE
Refactor TE-14.1 with updated scale params

### PR DESCRIPTION
**Pull request by self-maurya**
default scale params

- V4TunnelCount = 20000
- V4TunnelNHGCount = 256
- V4TunnelNHGSplitCount = 2
- EgressNHGSplitCount  = 16
- V4ReEncapNHGCount = 256

### CODE REVIEW

File: internal/tescale/scale.go [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+	V4TunnelNHGCount      int
+	V4TunnelNHGSplitCount int
+	EgressNHGSplitCount   int
+	V4ReEncapNHGCount     int
+}
+
+// BuildVRFConfig creates scale new scale VRF configurations.
+func BuildVRFConfig(dut *ondatra.DUTDevice, egressIPs []string, param Param) []*VRFConfig {
+	v4TunnelIPAddrs := NewIPPool(iputil.GenerateIPs(V4TunnelIPBlock, param.V4TunnelCount))
+	v4VIPAddrs := NewIPPool(iputil.GenerateIPs(V4VIPIPBlock, (param.V4TunnelNHGCount*param.V4TunnelNHGSplitCount)+2))
```
Comment (prakashbadri-arista) [inline internal/tescale/scale.go:125]: Looks like we are creating over 500 VIP routes. in prod, the expectation is that there will only be 2 VIP routes in default VRF.

File: internal/tescale/scale.go [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+			nhgID = idPool.NextNHGID()
+
+			// Build NHs and link NHs to NHG.
+			for i := 0; i < param.V4TunnelNHGSplitCount; i++ {
+				vip := v4VIPAddrs.NextIP()
+				nhID = idPool.NextNHID()
+				vrfDefault.NHs = append(vrfDefault.NHs,
+					fluent.NextHopEntry().WithIndex(nhID).WithNetworkInstance(defaultVRF).WithIPAddress(vip),
+				)
+				vrfDefault.NHGs = append(vrfDefault.NHGs,
```
Comment (prakashbadri-arista) [inline internal/tescale/scale.go:164]: why are we appending the same NHG ID multiple times.  Shouldnt the NHs be created inside the 2nd for loop and the NHG added to "vrfDefault.NHGs" once outside with the no. of nhs = param.V4TunnelNHGSplitCount

File: internal/tescale/scale.go [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+		nhgID := idPool.NextNHGID()
+
+		// Build NHs and link NHs to NHG.
+		for i := 0; i < param.EgressNHGSplitCount; i++ {
+			vip := v4EgressIPAddrs.AllIPs()[i]
+			nhID = idPool.NextNHID()
+			vrfDefault.NHs = append(vrfDefault.NHs,
+				fluent.NextHopEntry().WithIndex(nhID).WithNetworkInstance(defaultVRF).WithIPAddress(vip),
+			)
+			vrfDefault.NHGs = append(vrfDefault.NHGs,
```
Comment (prakashbadri-arista) [inline internal/tescale/scale.go:190]: same comment as earlier..  the nhg is appended multiple times to vrfDefault.NHGs.

File: internal/tescale/scale.go [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+				)
+			}
+		}
+
+		// Build IPv4 entry
+		vrfRDConf.NHs = append(vrfRDConf.NHs,
+			fluent.IPv4Entry().WithPrefix(ip).WithNextHopGroup(nhgID).WithNetworkInstance(VRFRD).WithNextHopGroupNetworkInstance(defaultVRF),
+		)
+	}
+
```
Comment (prakashbadri-arista) [inline internal/tescale/scale.go:265]: can you add a log printing the no of routes in each vrf, no. of nhgs, no. of nhs created. 

something like
"""
   t.Logf("Created nhs: %v ", len(vrfDefault.NHs))
   t.Logf("Created nhgs: %v ", len(vrfDefault.NHGs))
   t.Logf("Created v4Ent: %v ", len(vrfRConf.V4Entries))
"""

Will be easier to debug

File: internal/tescale/scale.go [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+		if idx%reEncapNHGRatio == 0 {
+			nhgID = idPool.NextNHGID()
+		}
+
+		nhID = idPool.NextNHID()
+		vrfDefault.NHs = append(vrfDefault.NHs,
+			fluent.NextHopEntry().WithIndex(nhID).WithDecapsulateHeader(fluent.IPinIP).WithEncapsulateHeader(fluent.IPinIP).
+				WithNetworkInstance(defaultVRF).WithIPinIP(tunnelSrcIP, v4TunnelIPAddrs.AllIPs()[(idx+1)%len(v4TunnelIPAddrs.AllIPs())]),
+		)
+		vrfDefault.NHGs = append(vrfDefault.NHGs,
```
Comment (prakashbadri-arista) [inline internal/tescale/scale.go:229]: add to the list only when a new NHG id is allocated and created. This causing multiple ADD requests for the same NHG to the server

**Review by prakashbadri-arista — COMMENTED** [review]
Hi Pramod,
Can you add details regarding  the number of routes, NHGs and NHs and no. of NHs per NHG created in the scale test  as part of the PR description.

File: internal/tescale/scale.go (hunk 1 above)
Comment (self-maurya) [reply internal/tescale/scale.go:125]: @xw-g could you please confirm check and confirm

File: internal/tescale/scale.go [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+					fluent.NextHopEntry().WithIndex(nhID).WithNetworkInstance(defaultVRF).WithIPAddress(vip),
+				)
+				nhgEntry = nhgEntry.AddNextHop(nhID, 1)
+			}
+			vrfDefault.NHGs = append(vrfDefault.NHGs, nhgEntry)
+		}
+
+		// Build IPv4 entry
+		vrfTConf.V4Entries = append(vrfTConf.V4Entries,
+			fluent.IPv4Entry().WithPrefix(ip).WithNextHopGroup(nhgID).WithNetworkInstance(VRFT).WithNextHopGroupNetworkInstance(defaultVRF),
```
Comment (prakashbadri-arista) [inline internal/tescale/scale.go:172]: it should be fluent.IPv4Entry().WithPrefix(ip + "/32" ).

The client is publishing prefixes with no prefix length portion set and that is causing the failures.

File: internal/tescale/scale.go [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+				fluent.NextHopEntry().WithIndex(nhID).WithNetworkInstance(defaultVRF).WithIPAddress(vip),
+			)
+			nhgEntry = nhgEntry.AddNextHop(nhID, 1)
+		}
+
+		vrfDefault.NHGs = append(vrfDefault.NHGs, nhgEntry)
+		// Build IPv4 entry
+		vrfDefault.V4Entries = append(vrfDefault.V4Entries,
+			fluent.IPv4Entry().WithPrefix(ip).WithNextHopGroup(nhgID).WithNetworkInstance(defaultVRF).WithNextHopGroupNetworkInstance(defaultVRF),
+		)
```
Comment (prakashbadri-arista) [inline internal/tescale/scale.go:197]: please set the prefix ( /32 ) for the routes here and other places where Ipv4Entry is created

**Review by rohit-rp — APPROVED** [review]

**Pull request merged**